File: sdk/go/webhook_test.go
package iapstack

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"io"
	"net/http"
	"net/http/httptest"
	"strconv"
	"strings"
	"testing"
	"time"
)

const (
	// testWebhookSecret is a non-production HMAC fixture with the required entropy length.
	testWebhookSecret = "0123456789abcdef0123456789abcdef"
	// testWebhookBody is a production-shaped entitlement.changed payload.
	testWebhookBody = `{` +
		`"schema_version":1,` +
		`"project_id":"project-1",` +
		`"application_id":"ios-sandbox",` +
		`"customer_id":"customer-internal",` +
		`"entitlement_id":"entitlement-1",` +
		`"entitlement_key":"premium",` +
		`"access":"allowed",` +
		`"access_reason":"purchase_valid",` +
		`"source_observation_id":"observation-1",` +
		`"source_application_id":"ios-sandbox",` +
		`"source_product_id":"premium_annual",` +
		`"version":1` +
		`}`
)

// TestWebhookVerifierAcceptsAuthenticatedEvent verifies exact v2 signature validation.
func TestWebhookVerifierAcceptsAuthenticatedEvent(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, time.August, 31, 10, 0, 0, 0, time.UTC)
	verifier := newTestVerifier(t, now, defaultWebhookBodyLimit)
	defer verifier.Close()

	event, err := verifier.VerifyRequest(signedWebhookRequest(t, now, []byte(testWebhookBody), "event-1"))
	if err != nil {
		t.Fatalf("VerifyRequest() error = %v", err)
	}
	if event.ID != "event-1" || event.Duplicate || event.Change.EntitlementKey != "premium" {
		t.Fatalf("event = %#v", event)
	}
	if !event.Change.Entitlement().GrantsAccessAt(now) {
		t.Fatal("authenticated entitlement change should grant access")
	}
}

// TestWebhookVerifierRejectsInvalidSignature verifies unauthenticated bodies never reach storage.
func TestWebhookVerifierRejectsInvalidSignature(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, time.August, 31, 10, 0, 0, 0, time.UTC)
	store := NewMemoryEventStore()
	verifier := newTestVerifierWithStore(t, now, defaultWebhookBodyLimit, store)
	defer verifier.Close()

	request := signedWebhookRequest(t, now, []byte(testWebhookBody), "event-1")
	request.Header.Set(headerSignature, signaturePrefix+hex.EncodeToString(make([]byte, sha256.Size)))
	if _, err := verifier.VerifyRequest(request); !webhookCode(err, "invalid_signature") {
		t.Fatalf("error = %v, want invalid_signature", err)
	}
	if duplicate, err := store.Remember(request.Context(), "event-1", sha256.Sum256([]byte(testWebhookBody))); err != nil || duplicate {
		t.Fatalf("store remembered a rejected delivery: duplicate=%t err=%v", duplicate, err)
	}
}

// TestWebhookVerifierRejectsStaleTimestamp verifies replay-window enforcement precedes persistence.
func TestWebhookVerifierRejectsStaleTimestamp(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, time.August, 31, 10, 0, 0, 0, time.UTC)
	verifier := newTestVerifier(t, now, defaultWebhookBodyLimit)
	defer verifier.Close()

	request := signedWebhookRequest(t, now.Add(-6*time.Minute), []byte(testWebhookBody), "event-1")
	if _, err := verifier.VerifyRequest(request); !webhookCode(err, "invalid_timestamp") {
		t.Fatalf("error = %v, want invalid_timestamp", err)
	}
}

// TestWebhookVerifierRejectsFutureTimestampOutsideWindow verifies clock-skew bounds.
func TestWebhookVerifierRejectsFutureTimestampOutsideWindow(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, time.August, 31, 10, 0, 0, 0, time.UTC)
	verifier := newTestVerifier(t, now, defaultWebhookBodyLimit)
	defer verifier.Close()

	request := signedWebhookRequest(t, now.Add(6*time.Minute), []byte(testWebhookBody), "event-1")
	if _, err := verifier.VerifyRequest(request); !webhookCode(err, "invalid_timestamp") {
		t.Fatalf("error = %v, want invalid_timestamp", err)
	}
}

// TestWebhookVerifierRejectsEventIDSubstitution verifies the event ID is inside the MAC.
func TestWebhookVerifierRejectsEventIDSubstitution(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, time.August, 31, 10, 0, 0, 0, time.UTC)
	verifier := newTestVerifier(t, now, defaultWebhookBodyLimit)
	defer verifier.Close()

	request := signedWebhookRequest(t, now, []byte(testWebhookBody), "event-1")
	request.Header.Set(headerEventID, "attacker-new-event")
	if _, err := verifier.VerifyRequest(request); !webhookCode(err, "invalid_signature") {
		t.Fatalf("error = %v, want invalid_signature", err)
	}
}

// TestWebhookVerifierRejectsLegacyV1Signatures verifies v1 MACs cannot authenticate a delivery.
func TestWebhookVerifierRejectsLegacyV1Signatures(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, time.August, 31, 10, 0, 0, 0, time.UTC)
	verifier := newTestVerifier(t, now, defaultWebhookBodyLimit)
	defer verifier.Close()

	body := []byte(testWebhookBody)
	timestampText := strconv.FormatInt(now.Unix(), 10)
	mac := hmac.New(sha256.New, []byte(testWebhookSecret))
	_, _ = io.WriteString(mac, timestampText+".")
	_, _ = mac.Write(body)
	request := httptest.NewRequest(http.MethodPost, "/webhooks/iapstack", bytes.NewReader(body))
	request.Header.Set("Content-Type", jsonContentType)
	request.Header.Set(headerEventID, "attacker-new-event")
	request.Header.Set(headerTimestamp, timestampText)
	request.Header.Set(headerSignature, "v1="+hex.EncodeToString(mac.Sum(nil)))
	if _, err := verifier.VerifyRequest(request); !webhookCode(err, "invalid_signature") {
		t.Fatalf("error = %v, want invalid_signature", err)
	}
}

// TestWebhookVerifierDeduplicatesExactAuthenticatedReplay verifies one logical event ID.
func TestWebhookVerifierDeduplicatesExactAuthenticatedReplay(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, time.August, 31, 10, 0, 0, 0, time.UTC)
	verifier := newTestVerifier(t, now, defaultWebhookBodyLimit)
	defer verifier.Close()

	first, err := verifier.VerifyRequest(signedWebhookRequest(t, now, []byte(testWebhookBody), "event-1"))
	if err != nil || first.Duplicate {
		t.Fatalf("first delivery = %#v err=%v", first, err)
	}
	second, err := verifier.VerifyRequest(signedWebhookRequest(t, now, []byte(testWebhookBody), "event-1"))
	if err != nil || !second.Duplicate {
		t.Fatalf("replay = %#v err=%v", second, err)
	}
}

// TestWebhookVerifierRejectsEventIdentityConflict verifies altered replays fail closed.
func TestWebhookVerifierRejectsEventIdentityConflict(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, time.August, 31, 10, 0, 0, 0, time.UTC)
	verifier := newTestVerifier(t, now, defaultWebhookBodyLimit)
	defer verifier.Close()

	if _, err := verifier.VerifyRequest(signedWebhookRequest(t, now, []byte(testWebhookBody), "event-1")); err != nil {
		t.Fatalf("first delivery error = %v", err)
	}
	altered := []byte(strings.Replace(testWebhookBody, `"version":1`, `"version":2`, 1))
	if _, err := verifier.VerifyRequest(signedWebhookRequest(t, now, altered, "event-1")); !webhookCode(err, "event_identity_conflict") {
		t.Fatalf("error = %v, want event_identity_conflict", err)
	}
}

// TestWebhookVerifierBoundsWebhookBody verifies oversized input is rejected before HMAC work.
func TestWebhookVerifierBoundsWebhookBody(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, time.August, 31, 10, 0, 0, 0, time.UTC)
	verifier := newTestVerifier(t, now, 16)
	defer verifier.Close()

	body := bytes.Repeat([]byte("a"), 17)
	if _, err := verifier.VerifyRequest(signedWebhookRequest(t, now, body, "event-1")); !webhookCode(err, "body_too_large") {
		t.Fatalf("error = %v, want body_too_large", err)
	}
}

// TestNewWebhookVerifierRejectsShortSecrets verifies the HMAC key length boundary.
func TestNewWebhookVerifierRejectsShortSecrets(t *testing.T) {
	t.Parallel()

	_, err := NewWebhookVerifier(WebhookConfig{Secret: []byte("short"), Store: NewMemoryEventStore()})
	if err == nil {
		t.Fatal("NewWebhookVerifier() accepted a short secret")
	}
}

// TestWebhookVerifierRejectsMethodAndContentType verifies request envelope checks.
func TestWebhookVerifierRejectsMethodAndContentType(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, time.August, 31, 10, 0, 0, 0, time.UTC)
	verifier := newTestVerifier(t, now, defaultWebhookBodyLimit)
	defer verifier.Close()

	getRequest := httptest.NewRequest(http.MethodGet, "/webhooks/iapstack", bytes.NewReader([]byte(testWebhookBody)))
	if _, err := verifier.VerifyRequest(getRequest); !webhookCode(err, "method_not_allowed") {
		t.Fatalf("GET error = %v, want method_not_allowed", err)
	}
	request := signedWebhookRequest(t, now, []byte(testWebhookBody), "event-1")
	request.Header.Set("Content-Type", "text/plain")
	if _, err := verifier.VerifyRequest(request); !webhookCode(err, "content_type_required") {
		t.Fatalf("content type error = %v, want content_type_required", err)
	}
}

// TestWebhookVerifierRejectsInvalidEventIDAndPayload verifies identity and body contracts.
func TestWebhookVerifierRejectsInvalidEventIDAndPayload(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, time.August, 31, 10, 0, 0, 0, time.UTC)
	verifier := newTestVerifier(t, now, defaultWebhookBodyLimit)
	defer verifier.Close()

	invalidID := signedWebhookRequest(t, now, []byte(testWebhookBody), "event 1")
	if _, err := verifier.VerifyRequest(invalidID); !webhookCode(err, "invalid_event_id") {
		t.Fatalf("event ID error = %v, want invalid_event_id", err)
	}
	invalidJSON := signedWebhookRequest(t, now, []byte(`{"schema_version":1}`), "event-1")
	if _, err := verifier.VerifyRequest(invalidJSON); !webhookCode(err, "invalid_event") {
		t.Fatalf("payload error = %v, want invalid_event", err)
	}
}

// TestWebhookVerifierAppliesDefaultBounds verifies zero config values become production defaults.
func TestWebhookVerifierAppliesDefaultBounds(t *testing.T) {
	t.Parallel()

	now := time.Date(2026, time.August, 31, 10, 0, 0, 0, time.UTC)
	verifier, err := NewWebhookVerifier(WebhookConfig{
		Secret: []byte(testWebhookSecret),
		Store:  NewMemoryEventStore(),
		Clock:  func() time.Time { return now },
	})
	if err != nil {
		t.Fatalf("NewWebhookVerifier() error = %v", err)
	}
	defer verifier.Close()
	if _, err := verifier.VerifyRequest(signedWebhookRequest(t, now, []byte(testWebhookBody), "event-1")); err != nil {
		t.Fatalf("VerifyRequest() error = %v", err)
	}
}

// TestWebhookVerifierRejectsNilRequest verifies a missing HTTP request fails closed.
func TestWebhookVerifierRejectsNilRequest(t *testing.T) {
	t.Parallel()

	verifier := newTestVerifier(t, time.Now().UTC(), defaultWebhookBodyLimit)
	defer verifier.Close()
	if _, err := verifier.VerifyRequest(nil); !webhookCode(err, "invalid_body") {
		t.Fatalf("error = %v, want invalid_body", err)
	}
}

// newTestVerifier constructs a verifier with an isolated memory store.
func newTestVerifier(t *testing.T, now time.Time, bodyLimit int64) *WebhookVerifier {
	t.Helper()
	return newTestVerifierWithStore(t, now, bodyLimit, NewMemoryEventStore())
}

// newTestVerifierWithStore constructs a verifier around one injected store.
func newTestVerifierWithStore(t *testing.T, now time.Time, bodyLimit int64, store EventStore) *WebhookVerifier {
	t.Helper()

	verifier, err := NewWebhookVerifier(WebhookConfig{
		Secret:             []byte(testWebhookSecret),
		BodyLimit:          bodyLimit,
		TimestampTolerance: 5 * time.Minute,
		Store:              store,
		Clock:              func() time.Time { return now },
	})
	if err != nil {
		t.Fatalf("NewWebhookVerifier() error = %v", err)
	}
	return verifier
}

// signedWebhookRequest builds one production-compatible v2 delivery.
func signedWebhookRequest(t *testing.T, timestamp time.Time, body []byte, eventID string) *http.Request {
	t.Helper()

	request := httptest.NewRequest(http.MethodPost, "/webhooks/iapstack", bytes.NewReader(body))
	request.Header.Set("Content-Type", jsonContentType)
	request.Header.Set(headerEventID, eventID)
	request.Header.Set(headerTimestamp, strconv.FormatInt(timestamp.Unix(), 10))
	request.Header.Set(headerSignature, "v2="+signWebhook(eventID, timestamp, body))
	return request
}

// signWebhook computes HMAC-SHA-256 over the documented newline-delimited v2 input.
func signWebhook(eventID string, timestamp time.Time, body []byte) string {
	mac := hmac.New(sha256.New, []byte(testWebhookSecret))
	_, _ = io.WriteString(mac, signatureVersion)
	_, _ = io.WriteString(mac, signatureMACSeparator)
	_, _ = io.WriteString(mac, eventID)
	_, _ = io.WriteString(mac, signatureMACSeparator)
	_, _ = io.WriteString(mac, strconv.FormatInt(timestamp.Unix(), 10))
	_, _ = io.WriteString(mac, signatureMACSeparator)
	_, _ = mac.Write(body)
	return hex.EncodeToString(mac.Sum(nil))
}

// webhookCode reports whether err is a WebhookError with the expected code.
func webhookCode(err error, code string) bool {
	var webhookErr *WebhookError
	return errors.As(err, &webhookErr) && webhookErr.Code == code
}

File: sdk/typescript/test/webhook.test.ts
import { test } from 'bun:test';
import assert from 'node:assert/strict';
import { createHash, createHmac } from 'node:crypto';
import { webhookStatus } from '../src/errors';
import {
  MemoryEventStore,
  WebhookError,
  WebhookVerifier,
  type EventStore,
  type WebhookRequest,
} from '../src/index';

const testWebhookSecret = '0123456789abcdef0123456789abcdef';
const testWebhookBody =
  '{' +
  '"schema_version":1,' +
  '"project_id":"project-1",' +
  '"application_id":"ios-sandbox",' +
  '"customer_id":"customer-internal",' +
  '"entitlement_id":"entitlement-1",' +
  '"entitlement_key":"premium",' +
  '"access":"allowed",' +
  '"access_reason":"purchase_valid",' +
  '"source_observation_id":"observation-1",' +
  '"source_application_id":"ios-sandbox",' +
  '"source_product_id":"premium_annual",' +
  '"version":1' +
  '}';

const defaultWebhookBodyLimit = 1 << 20;

class FailingStore implements EventStore {
  constructor(private readonly error: unknown) {}

  remember(): never {
    throw this.error;
  }
}

function webhookCode(error: unknown, code: string): boolean {
  return error instanceof WebhookError && error.code === code && error.statusCode === webhookStatus(code);
}

function signWebhook(eventId: string, timestamp: Date, body: Uint8Array | string): string {
  const mac = createHmac('sha256', testWebhookSecret);
  mac.update('v2');
  mac.update('\n');
  mac.update(eventId);
  mac.update('\n');
  mac.update(String(Math.floor(timestamp.getTime() / 1000)));
  mac.update('\n');
  mac.update(body);
  return mac.digest('hex');
}

function signedWebhookRequest(
  timestamp: Date,
  body: Uint8Array | string,
  eventId: string,
): WebhookRequest {
  const payload = typeof body === 'string' ? Buffer.from(body) : body;
  return {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'IAPStack-Event-ID': eventId,
      'IAPStack-Timestamp': String(Math.floor(timestamp.getTime() / 1000)),
      'IAPStack-Signature': `v2=${signWebhook(eventId, timestamp, payload)}`,
    },
    body: payload,
  };
}

function newTestVerifier(
  now: Date,
  bodyLimit = defaultWebhookBodyLimit,
  store = new MemoryEventStore(),
): WebhookVerifier {
  return new WebhookVerifier({
    secret: testWebhookSecret,
    bodyLimit,
    timestampToleranceMs: 5 * 60 * 1000,
    store,
    clock: () => now,
  });
}

test('webhook verifier accepts an authenticated event', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now);
  const event = await verifier.verifyRequest(signedWebhookRequest(now, testWebhookBody, 'event-1'));
  assert.equal(event.id, 'event-1');
  assert.equal(event.duplicate, false);
  assert.equal(event.change.entitlementKey, 'premium');
  assert.equal(event.change.entitlement().grantsAccessAt(now), true);
  verifier.close();
});

test('webhook verifier rejects an invalid signature before storage', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const store = new MemoryEventStore();
  const verifier = newTestVerifier(now, defaultWebhookBodyLimit, store);
  const request = signedWebhookRequest(now, testWebhookBody, 'event-1');
  request.headers = {
    ...request.headers,
    'IAPStack-Signature': `v2=${'00'.repeat(32)}`,
  };
  await assert.rejects(
    () => verifier.verifyRequest(request),
    (error) => webhookCode(error, 'invalid_signature'),
  );
  assert.equal(store.remember('event-1', createHash('sha256').update(testWebhookBody).digest()), false);
  verifier.close();
});

test('webhook verifier rejects a stale timestamp', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now);
  const request = signedWebhookRequest(
    new Date(now.getTime() - 6 * 60 * 1000),
    testWebhookBody,
    'event-1',
  );
  await assert.rejects(
    () => verifier.verifyRequest(request),
    (error) => webhookCode(error, 'invalid_timestamp'),
  );
  verifier.close();
});

test('webhook verifier rejects a future timestamp outside the window', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now);
  const request = signedWebhookRequest(
    new Date(now.getTime() + 6 * 60 * 1000),
    testWebhookBody,
    'event-1',
  );
  await assert.rejects(
    () => verifier.verifyRequest(request),
    (error) => webhookCode(error, 'invalid_timestamp'),
  );
  verifier.close();
});

test('webhook verifier rejects event-ID substitution', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now);
  const request = signedWebhookRequest(now, testWebhookBody, 'event-1');
  request.headers = { ...request.headers, 'IAPStack-Event-ID': 'attacker-new-event' };
  await assert.rejects(
    () => verifier.verifyRequest(request),
    (error) => webhookCode(error, 'invalid_signature'),
  );
  verifier.close();
});

test('webhook verifier rejects legacy v1 signatures', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now);
  const timestampText = String(Math.floor(now.getTime() / 1000));
  const mac = createHmac('sha256', testWebhookSecret);
  mac.update(`${timestampText}.`);
  mac.update(testWebhookBody);
  await assert.rejects(
    () =>
      verifier.verifyRequest({
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'IAPStack-Event-ID': 'attacker-new-event',
          'IAPStack-Timestamp': timestampText,
          'IAPStack-Signature': `v1=${mac.digest('hex')}`,
        },
        body: testWebhookBody,
      }),
    (error) => webhookCode(error, 'invalid_signature'),
  );
  verifier.close();
});

test('webhook verifier deduplicates an exact authenticated replay', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now);
  const first = await verifier.verifyRequest(signedWebhookRequest(now, testWebhookBody, 'event-1'));
  const second = await verifier.verifyRequest(signedWebhookRequest(now, testWebhookBody, 'event-1'));
  assert.equal(first.duplicate, false);
  assert.equal(second.duplicate, true);
  verifier.close();
});

test('webhook verifier rejects an event identity conflict', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now);
  await verifier.verifyRequest(signedWebhookRequest(now, testWebhookBody, 'event-1'));
  const altered = testWebhookBody.replace('"version":1', '"version":2');
  await assert.rejects(
    () => verifier.verifyRequest(signedWebhookRequest(now, altered, 'event-1')),
    (error) => webhookCode(error, 'event_identity_conflict'),
  );
  verifier.close();
});

test('webhook verifier bounds the webhook body', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now, 16);
  const body = 'a'.repeat(17);
  await assert.rejects(
    () => verifier.verifyRequest(signedWebhookRequest(now, body, 'event-1')),
    (error) => webhookCode(error, 'body_too_large'),
  );
  verifier.close();
});

test('NewWebhookVerifier rejects a short secret', () => {
  assert.throws(
    () => new WebhookVerifier({ secret: 'short', store: new MemoryEventStore() }),
  );
});

test('webhook verifier rejects method and content type', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now);
  await assert.rejects(
    () =>
      verifier.verifyRequest({
        method: 'GET',
        headers: { 'Content-Type': 'application/json' },
        body: testWebhookBody,
      }),
    (error) => webhookCode(error, 'method_not_allowed'),
  );
  const request = signedWebhookRequest(now, testWebhookBody, 'event-1');
  request.headers = { ...request.headers, 'Content-Type': 'text/plain' };
  await assert.rejects(
    () => verifier.verifyRequest(request),
    (error) => webhookCode(error, 'content_type_required'),
  );
  verifier.close();
});

test('webhook verifier rejects invalid event IDs and payloads', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now);
  await assert.rejects(
    () => verifier.verifyRequest(signedWebhookRequest(now, testWebhookBody, 'event 1')),
    (error) => webhookCode(error, 'invalid_event_id'),
  );
  await assert.rejects(
    () => verifier.verifyRequest(signedWebhookRequest(now, '{"schema_version":1}', 'event-1')),
    (error) => webhookCode(error, 'invalid_event'),
  );
  verifier.close();
});

test('webhook verifier applies default bounds', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = new WebhookVerifier({
    secret: testWebhookSecret,
    store: new MemoryEventStore(),
    clock: () => now,
  });
  await verifier.verifyRequest(signedWebhookRequest(now, testWebhookBody, 'event-1'));
  verifier.close();
});

test('webhook verifier rejects a missing request', async () => {
  const verifier = newTestVerifier(new Date());
  await assert.rejects(
    () => verifier.verifyRequest(null),
    (error) => webhookCode(error, 'invalid_body'),
  );
  verifier.close();
});

test('webhook verifier accepts Node lowercase headers and JSON charset', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now);
  const eventId = 'event-node';
  const body = Buffer.from(testWebhookBody);
  const event = await verifier.verifyRequest({
    method: 'POST',
    headers: {
      'content-type': 'application/json; charset=utf-8',
      'iapstack-event-id': eventId,
      'iapstack-timestamp': String(Math.floor(now.getTime() / 1000)),
      'iapstack-signature': `v2=${signWebhook(eventId, now, body)}`,
    },
    body,
  });
  assert.equal(event.id, eventId);
  verifier.close();
});

test('webhook verifier accepts a Fetch Request', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now);
  const unsigned = signedWebhookRequest(now, testWebhookBody, 'event-fetch');
  const headers = new Headers(unsigned.headers as Record<string, string>);
  const request = new Request('https://host.example/webhooks/iapstack', {
    method: 'POST',
    headers,
    body: testWebhookBody,
  });
  const event = await verifier.verifyRequest(request);
  assert.equal(event.id, 'event-fetch');
  verifier.close();
});

test('webhook verifier wraps store failures as unavailable', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const cause = new Error('disk full');
  const verifier = newTestVerifier(now, defaultWebhookBodyLimit, new FailingStore(cause));
  await assert.rejects(
    () => verifier.verifyRequest(signedWebhookRequest(now, testWebhookBody, 'event-1')),
    (error: unknown) => {
      assert.equal(webhookCode(error, 'receiver_unavailable'), true);
      assert(error instanceof WebhookError);
      assert.equal(error.cause, cause);
      return true;
    },
  );
  verifier.close();

  const conflictVerifier = newTestVerifier(
    now,
    defaultWebhookBodyLimit,
    new FailingStore(new WebhookError('event_identity_conflict')),
  );
  await assert.rejects(
    () => conflictVerifier.verifyRequest(signedWebhookRequest(now, testWebhookBody, 'event-2')),
    (error) => webhookCode(error, 'event_identity_conflict'),
  );
  conflictVerifier.close();
});

test('webhook handler owns IAPStack retry statuses', async () => {
  const now = new Date('2026-08-31T10:00:00.000Z');
  const verifier = newTestVerifier(now);
  const handle = verifier.handler(async () => undefined);

  const accepted = await handle(signedWebhookRequest(now, testWebhookBody, 'event-1'));
  assert.equal(accepted.status, 204);

  const duplicate = await handle(signedWebhookRequest(now, testWebhookBody, 'event-1'));
  assert.equal(duplicate.status, 204);

  const badSignature = signedWebhookRequest(now, testWebhookBody, 'event-2');
  badSignature.headers = {
    ...badSignature.headers,
    'IAPStack-Signature': `v2=${'00'.repeat(32)}`,
  };
  const unauthorized = await handle(badSignature);
  assert.equal(unauthorized.status, 401);
  assert.deepEqual(await unauthorized.json(), { code: 'invalid_signature' });

  const conflictVerifier = newTestVerifier(now);
  await conflictVerifier.verifyRequest(signedWebhookRequest(now, testWebhookBody, 'event-3'));
  const altered = testWebhookBody.replace('"version":1', '"version":2');
  const conflict = await conflictVerifier.handler(async () => undefined)(
    signedWebhookRequest(now, altered, 'event-3'),
  );
  assert.equal(conflict.status, 409);
  assert.deepEqual(await conflict.json(), { code: 'event_identity_conflict' });
  conflictVerifier.close();

  const unavailableVerifier = newTestVerifier(
    now,
    defaultWebhookBodyLimit,
    new FailingStore(new Error('disk full')),
  );
  const unavailable = await unavailableVerifier.handler(async () => undefined)(
    signedWebhookRequest(now, testWebhookBody, 'event-4'),
  );
  assert.equal(unavailable.status, 503);
  assert.deepEqual(await unavailable.json(), { code: 'receiver_unavailable' });
  unavailableVerifier.close();

  const callbackVerifier = newTestVerifier(now);
  const callbackFailed = await callbackVerifier.handler(async () => {
    throw new Error('host side effect failed');
  })(signedWebhookRequest(now, testWebhookBody, 'event-5'));
  assert.equal(callbackFailed.status, 503);
  assert.deepEqual(await callbackFailed.json(), { code: 'receiver_unavailable' });
  callbackVerifier.close();

  const method = await handle({
    method: 'GET',
    headers: { 'Content-Type': 'application/json' },
    body: testWebhookBody,
  });
  assert.equal(method.status, 405);
  assert.equal(method.headers.get('Allow'), 'POST');

  const noop = await verifier.handler()(signedWebhookRequest(now, testWebhookBody, 'event-6'));
  assert.equal(noop.status, 204);

  verifier.close();
});
